src/log/Log.cc

```diff
@@ -239,8 +278,52 @@ void Log::_flush(EntryQueue *t, EntryQueue *requeue, bool crash)
       if (do_stderr) {
 	cerr << buf << s << std::endl;
       }
-    }
 
+      if (do_graylog2) {
+	std::auto_ptr<Formatter> f(Formatter::create("json"));
+
+	char fsid_str[40];
+	m_fsid.print(fsid_str);
+
+	// GELF format
+	// http://www.graylog2.org/resources/gelf/specification
+	f->open_object_section("");
+	f->dump_string("version", "1.1");
+	f->dump_string("host", m_host);
+	f->dump_string("short_message", s);
+	f->dump_string("_app", "ceph");
+	f->dump_float("timestamp", e->m_stamp.sec() + (e->m_stamp.usec() / 1000000.0));
+	f->dump_int("_thread", e->m_thread);
+	f->dump_int("_prio", e->m_prio);
+	f->dump_string("_subsys_name", m_subs->get_name(sub));
+	f->dump_int("_subsys_id", sub);
+	f->dump_int("_crash", crash);
+	f->dump_string("_fsid", fsid_str);
+	if (crash) f->dump_int("_crash_entry_queue_len", -t->m_len);
+	f->close_section();
+
+	std::stringstream zos(std::stringstream::in |
+			      std::stringstream::out |
+			      std::stringstream::binary);
+
+	{
+	  boost::iostreams::filtering_ostream os;
+	  os.push(boost::iostreams::zlib_compressor());
+	  os.push(zos);
+
+	  f->flush(os);
+	  os << std::endl;
+	}
+
+	try {
+	  boost::asio::ip::udp::socket socket(m_graylog_io_service);
+	  socket.open(m_graylog_endpoint.protocol());
+	  socket.send_to(boost::asio::buffer(zos.str()), m_graylog_endpoint);
+	} catch (boost::system::system_error const& e) {
+	  /* The above code fails until the configuration is set */
+	}
+      }
+    }
```

Comment:
This looks functionally correct, but probably pretty slow relative to the other backends.  Formatter could be a static member of Log and re-used, along with fsid_str.  Maybe the zlib stream?  Not sure.

Anyway, that all can come later.